src/riak_core_throttle.erl

```diff
@@ -0,0 +1,301 @@
+%% Copyright (c) 2016 Basho Technologies, Inc.  All Rights Reserved.
+%%
+%% This file is provided to you under the Apache License,
+%% Version 2.0 (the "License"); you may not use this file
+%% except in compliance with the License.  You may obtain
+%% a copy of the License at
+%%
+%%   http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing,
+%% software distributed under the License is distributed on an
+%% "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+%% KIND, either express or implied.  See the License for the
+%% specific language governing permissions and limitations
+%% under the License.
+%%
+%% -------------------------------------------------------------------
+%%
+%% @doc Throttling for activities that can potentially overload resources.
+%% Provides support for "throttling" (i.e. slowing down) calling processes
+%% that, when under heavy load, could potentially overload other resources.
+%% Activities are identified by a unique key so that different throttle values
+%% can be used for different activities.
+%%
+%% The throttle for an activity can be set to a specific value with the
+%% `set_throttle/2' function, or it can be set to a value based on the current
+%% "load" for the activity by using the `set_limits/2' function to define a
+%% mapping from load factors to throttle values, and then periodically calling
+%% the `set_throttle_by_load/2' function with the current load factor. Note
+%% that "load factor" is an abstract concept, the purpose of which is simply
+%% to represent numerically the amount of load that an activity is under and
+%% map that level of load to a particular throttle level. Examples of metrics
+%% that could be used for load factors include the size of the mailbox for a
+%% particular process, or the number of messages per second being sent to an
+%% external service such as Solr.
+-module(riak_core_throttle).
+
+%% API
+-export([get_throttle/1,
+         set_throttle/2,
+         clear_throttle/1,
+         disable_throttle/1,
+         enable_throttle/1,
+         is_throttle_enabled/1,
+         init/4,
+         set_limits/2,
+         get_limits/1,
+         clear_limits/1,
+         create_limits_translator_fun/2,
+         set_throttle_by_load/2,
+         throttle/1]).
+
+-ifdef(TEST).
+-export([get_throttle_for_load/2]).
+-define(SLEEP(Time), Time).
+-else.
+-define(SLEEP(Time), timer:sleep(Time)).
+-endif.
+
+-type activity_key() :: atom().
+-type throttle_time() :: non_neg_integer().
+-type load_factor() :: number() | atom() | tuple().
+-type limits() :: [{load_factor(), throttle_time()}].
+
+-export_type([activity_key/0, throttle_time/0, load_factor/0, limits/0]).
+
+-define(THROTTLE_KEY(Key), list_to_atom("throttle_" ++ atom_to_list(Key))).
+-define(THROTTLE_LIMITS_KEY(Key),
+        list_to_atom("throttle_limits_" ++ atom_to_list(Key))).
+-define(THROTTLE_KILL_KEY(Key),
```

Comment:
As discussed, we'll switch this from `THROTTLE_KILL_KEY` to `THROTTLE_ENABLED_KEY` and all the other things to not make this whole thing backwards, once everything else is working and tested.